testcases/misc_tests/obj_mgmt.c

```diff
@@ -1162,251 +1162,260 @@ CK_RV do_CreateTokenObjects(void)
+        {CKA_CLASS, &obj1_class,    sizeof(obj1_class)},
+        {CKA_TOKEN, &true,          sizeof(true)},
+        {CKA_LABEL, obj1_label,     sizeof(obj1_label)-1},
+        {CKA_VALUE, obj1_data,      sizeof(obj1_data)}
+    };
+
+    /* A secret key object */
+    CK_OBJECT_CLASS obj2_class = CKO_SECRET_KEY;
+    CK_KEY_TYPE     obj2_type = CKK_AES;
+    CK_UTF8CHAR     obj2_label[] = "Object 2";
+    CK_BYTE         obj2_data[AES_KEY_SIZE_128];
+    CK_ATTRIBUTE    obj2_template[] = {
+		{CKA_CLASS,     &obj2_class,    sizeof(obj2_class)},
+		{CKA_TOKEN,     &true,          sizeof(true)},
+		{CKA_KEY_TYPE,  &obj2_type,     sizeof(obj2_type)},
+		{CKA_LABEL,     obj2_label,     sizeof(obj2_label)-1},
+		{CKA_VALUE,     obj2_data,      sizeof(obj2_data)}
 	};
 
-        if (skip_token_obj == TRUE) {
-                testcase_notice("Skipping tests that creates token objects");
-                return CKR_OK;
+    CK_OBJECT_HANDLE h_counter1,
+                     h_clock,
+                     h_obj1,
+                     h_obj2,
+                     obj_list[10];
+
+    CK_ATTRIBUTE find_tmpl[] = {
+        {CKA_CLASS,	&counter1_class, sizeof(counter1_class)}
+    };
+
+    if (skip_token_obj == TRUE) {
+        testcase_notice("Skipping tests that creates token objects");
+        return CKR_OK;
+    }
+
+    slot_id = SLOT_ID;
+
+    testcase_begin("starting...");
+
+    if (get_user_pin(user_pin))
+        return CKR_FUNCTION_FAILED;
+
+    user_pin_len = (CK_ULONG)strlen((char *)user_pin);
+
+    /* Open a session with the token */
+    rc = funcs->C_OpenSession(slot_id,
+                              (CKF_SERIAL_SESSION|CKF_RW_SESSION),
+                              NULL_PTR,
+                              NULL_PTR,
+                              &h_session);
+    if (rc != CKR_OK) {
+        testcase_fail("C_OpenSession() rc = %s", p11_get_ckr(rc));
+        goto done;
+    }
+
+    // Login correctly
+    rc = funcs->C_Login(h_session, CKU_USER, user_pin, user_pin_len);
+    if (rc != CKR_OK) {
+        testcase_fail("C_Login() rc = %s", p11_get_ckr(rc));
+        goto session_close;
+    }
+
+    /* Create the 4 test objects */
+    rc = funcs->C_CreateObject(h_session, obj1_template, 4, &h_obj1);
+    if (rc != CKR_OK) {
+        testcase_fail("C_CreateObject() rc = %s", p11_get_ckr(rc));
+		return rc;
+	}
+
+    rc = funcs->C_CreateObject(h_session, obj2_template, 5, &h_obj2);
+    if (rc != CKR_OK) {
+        testcase_fail("C_CreateObject() rc = %s", p11_get_ckr(rc));
+        goto destroy_1;
+    }
+
+    /* try and create a monotonic object. This should fail
+     * since it is a read only feature.
+     */
+    rc = funcs->C_CreateObject(h_session, counter1_template, 6, &h_counter1);
+    if (rc != CKR_ATTRIBUTE_READ_ONLY) {
+        testcase_fail("C_CreateObject() rc = %s", p11_get_ckr(rc));
+        goto destroy_2;
+    }
+
+    rc = funcs->C_CreateObject(h_session, clock_template, 4, &h_clock);
+    if (rc != CKR_OK) {
```

Comment:
I don't really understand the hard coded numbers (5, 6, 4) in the calls to C_CreateObject.  I do see other testcases that also use similar numbers.  But I am not understanding how you know which ones to use.

Reply:
The third parameter that goes in a C_CreateObject call refers to the number of attributes in the template. For more information on this you can check on PKCS#11 standard.
In this case if you look at clock_template, it contains 4 attributes.